precompile/reward_manager.go

```diff
@@ -0,0 +1,472 @@
+	SetRewardAddressGasCost        uint64 = (writeGasCostPerSlot) + ReadAllowListGasCost // write 1 slot + read allow list
+
+	// RewardManagerRawABI contains the raw ABI of RewardManager contract.
+	RewardManagerRawABI = "[{\"inputs\":[],\"name\":\"allowFeeRecipients\",\"outputs\":[],\"stateMutability\":\"nonpayable\",\"type\":\"function\"},{\"inputs\":[],\"name\":\"areFeeRecipientsAllowed\",\"outputs\":[{\"internalType\":\"bool\",\"name\":\"isAllowed\",\"type\":\"bool\"}],\"stateMutability\":\"view\",\"type\":\"function\"},{\"inputs\":[],\"name\":\"currentRewardAddress\",\"outputs\":[{\"internalType\":\"address\",\"name\":\"rewardAddress\",\"type\":\"address\"}],\"stateMutability\":\"view\",\"type\":\"function\"},{\"inputs\":[],\"name\":\"disableRewards\",\"outputs\":[],\"stateMutability\":\"nonpayable\",\"type\":\"function\"},{\"inputs\":[{\"internalType\":\"address\",\"name\":\"addr\",\"type\":\"address\"}],\"name\":\"readAllowList\",\"outputs\":[{\"internalType\":\"uint256\",\"name\":\"role\",\"type\":\"uint256\"}],\"stateMutability\":\"view\",\"type\":\"function\"},{\"inputs\":[{\"internalType\":\"address\",\"name\":\"addr\",\"type\":\"address\"}],\"name\":\"setAdmin\",\"outputs\":[],\"stateMutability\":\"nonpayable\",\"type\":\"function\"},{\"inputs\":[{\"internalType\":\"address\",\"name\":\"addr\",\"type\":\"address\"}],\"name\":\"setEnabled\",\"outputs\":[],\"stateMutability\":\"nonpayable\",\"type\":\"function\"},{\"inputs\":[{\"internalType\":\"address\",\"name\":\"addr\",\"type\":\"address\"}],\"name\":\"setNone\",\"outputs\":[],\"stateMutability\":\"nonpayable\",\"type\":\"function\"},{\"inputs\":[{\"internalType\":\"address\",\"name\":\"addr\",\"type\":\"address\"}],\"name\":\"setRewardAddress\",\"outputs\":[],\"stateMutability\":\"nonpayable\",\"type\":\"function\"}]"
+)
+
+// Singleton StatefulPrecompiledContract and signatures.
+var (
+	_ StatefulPrecompileConfig = &RewardManagerConfig{}
+
+	ErrCannotAllowFeeRecipients      = errors.New("non-enabled cannot allowFeeRecipients")
+	ErrCannotAreFeeRecipientsAllowed = errors.New("non-enabled cannot areFeeRecipientsAllowed")
+	ErrCannotCurrentRewardAddress    = errors.New("non-enabled cannot currentRewardAddress")
+	ErrCannotDisableRewards          = errors.New("non-enabled cannot disableRewards")
+	ErrCannotSetRewardAddress        = errors.New("non-enabled cannot setRewardAddress")
+
+	ErrCannotEnableBothRewards = errors.New("cannot enable both fee recipients and reward address at the same time")
+	ErrEmptyRewardAddress      = errors.New("reward address cannot be empty")
+
+	RewardManagerABI        abi.ABI                     // will be initialized by init function
+	RewardManagerPrecompile StatefulPrecompiledContract // will be initialized by init function
+
+	rewardAddressStorageKey        = common.Hash{'r', 'a', 's', 'k'}
+	allowFeeRecipientsAddressValue = common.Address{'a', 'f', 'r', 'a', 'v'}
```

Comment:
Since this is an address it may collide with an address value supplied by a user accidentally, which seems like an unintended behavior.

Could we fix this by using a hash value here that populates the left most 12 bytes? Since an address converted to a hash takes up only the right most 20 bytes, this will guarantee that a user supplied address will not result in the same value.

Reply:
yep, makes utter sense. changed it to a reserved address.

Reply:
Can we switch to using a `common.Hash` value instead of a `common.Address` value instead? Since we left-pad the address values with zero bytes to create a 32 byte value, if we use a hash value that has a non-zero value in the left-most 12 bytes, then we will guarantee not to have any collisions without needing to reserve an additional address.